napari/components/viewer_model.py

```diff
@@ -995,6 +995,12 @@ def _add_layers_with_plugins(
             )
             # actually add the layer
             new = self._add_layer_from_data(*_data)
+            # store the source of the layer in metadata, if not present yet
+            if isinstance(new, list):
+                for layer in new:
+                    layer.metadata.setdefault('source', filename)
+            else:
+                new.metadata.setdefault('source', filename)
             # some add_* methods return a List[Layer], others just a Layer
             # we want to always return a list
             added.extend(new if isinstance(new, list) else [new])
```

Comment:
here is twice distinguishing between list and single element. Maybe simple:
```
if not isinstance(new, list):
	new = [new]
```
on begin?